src/bp/core/database/tables/bot-specific/messages.ts

```diff
@@ -0,0 +1,36 @@
+import { Table } from 'core/database/interfaces'
+
+export class MessagesTable extends Table {
+  readonly name: string = 'messages'
+
+  async bootstrap() {
+    let created = false
+
+    await this.knex.createTableIfNotExists(this.name, table => {
+      table.increments('id').primary()
+      table
+        .integer('conversationId')
+        .unsigned()
+        .references('id')
+        .inTable('conversations')
+        .notNullable()
+        .onDelete('cascade')
+      table.string('eventId')
+      /*
+        this whole thing can't work because the event is inserted after so the foreign key constraint fails
+        .references('id')
+        .inTable('events')
+        .nullable()
+        // not sure if this works
+        .onDelete('set null')
```

Comment:
I can't really make this work because the event inserts are batched so you can't know when inserting a message if the event actually exists in the db.

The only way to make this work would be to have some sort of standarized batching system with dependencies. So for example the batching for messages would be dependant on the batching for events. So the current batch of messages could only be inserted after the current batch of events has been inserted. 

It could still break if we have multiple nodes, but if we but batches to be max every 1-3 seconds the chances of breaking are very small.

This sort of system could work very well with the indexing (ids) system. Since we can't rely on the db to give use ids if we are batching inserts, we need to ask the indexing system to give us an id. This system could then be aware of which ids are currently in batches and which are not, and allow to query directly from memory for those that are before the insert happens

If we want to make this bulletproof, we could have each node of botpress keep a tab of what range of ids each node is currently giving out (let's say node 1 has reserved ids 10000-19999, node 2 20000-29999, node 30000-39999) and then if a query for a certain thing ends up not finding that thing in the db, it could check if the id was currently reserved by one of the nodes, asks this node to flush its batches, and try to run the query again

Reply:
Good idea but Let's not overthink this for now.